src/guards/slice_memory_guard.rs

```diff
@@ -50,6 +50,60 @@ impl<'a, T> SliceMemoryGuard<'a, T> {
         }
         SliceMemoryGuard { memory }
     }
+
+    /// Initialize memory guard using given iterator.
+    /// Automatically shrink's memory to given items' count.
+    /// `Ok(guard)` will be returned in this case.
+    ///
+    /// If items' count is too large to place in memory, moves it into new `Vec` and continue collecting into it.
+    /// `Err(vec)` will be returned in this case.
+    #[inline]
+    pub unsafe fn new_from_iter(memory: &'a mut [MaybeUninit<T>], source: impl Iterator<Item = T>) -> Result<Self, Vec<T>> {
+        let size_hint = source.size_hint();
+        let mut iter = source.peekable();
+
+        // Fulfilling placed memory
+        for (index, item) in memory.into_iter().enumerate() {
+            match iter.next() {
+                // While iterator returns new value, write it
+                Some(value) => {
+                    write(item.as_mut_ptr(), value);
+                },
+                // When it returns None then slicing memory and returning the guard
+                None => {
+                    return Ok(SliceMemoryGuard {
+                        memory: &mut memory[0..index],
+                    });
+                },
+            }
+        }
+
+        if iter.peek().is_some() {
+            // If iterator still contains values to return, collect it into the vector
+            let mut capacity_guess = usize::max(
+                // Blind guess
+                memory.len() + 1,
+                // Iterator lower hint
+                size_hint.0,
+            );
+            if let Some(iterator_upper_bound) = size_hint.1 {
+                capacity_guess = capacity_guess.max(iterator_upper_bound);
+            }
+            let mut vec = Vec::<T>::with_capacity(capacity_guess);
```

Comment:
what if we, instead of doing the guessing game, let Rust figure it out? i.e
```rust
let vec = memory
  .iter()
  .map(|item| read(item).assume_init())
  .chain(iter)
  .collect();
```

Reply:
I'd go with
```
let mut vec = Vec::new();
vec.extend_from_slice(memory.as_mut_ptr() as *mut T, memory.len());
vec.extend(iter);
```

Reply:
I suppose `chain` variant is better because of better `size_hint` impl for `Chain`

Reply:
This code is executed only when `size_hint` lied to us once already.
And `extend_from_slice` can be way faster than moving items from iterator one by one.

Reply:
Sadly, `extend_from_slice` and analogues require `Clone` to be implemented on `T`.

Reply:
`copy_nonoverlapping` did the same task done